### ISSUE TITLE
C10D performance?

### ISSUE THREAD

**Issue by vince62s** [pytorch/fairseq#298]
Hi,
Is there any benchmark of the impact of C10D vs no C10D ? 
thanks

**Comment by myleott** [comment]
We have one in the latest version of the Scaling NMT paper: https://arxiv.org/pdf/1806.00187.pdf. See Table 1, `16 nodes (from +2x lr)` vs. `+ overlap comm+bwd` (which is C10D). We see a ~19% increase in words per second.

**Comment by vince62s** [comment]
okay thanks, but I am confused. does C10D stand for Cuda 10 distributed ?
Because even with cuda 9, with pytorch 1.0 it imports C10D by default.

**Comment by myleott** [comment]
I don't know what c10d stands for, but it's the new shared distributed library for PyTorch and Caffe2. The main difference between the original implementation of DistributedDataParallel and the new c10d one is that the new one overlaps the backwards pass with communication.

**Comment by vince62s** [comment]
But I don't understand all the code related to C10D in Fairseq.
In my understanding, if C10D is installed then the __init__.py file in .../torch/distributed
automatically switch to the C10D library
if not installed then not.

Am I wrong ?

**Comment by myleott** [comment]
We need to support both the c10d and the old (deprecated) version of DistributedDataParallel, since the c10d version does not support graphs with conditional computation (e.g., language models with adaptive softmax), but the older deprecated one does.

We have all the extra checks so that we can support both older versions of pytorch (e.g., 0.4.1) and newer versions (e.g., current master).
